maitake/src/wait/dataqueue.rs

```diff
@@ -0,0 +1,984 @@
+#[repr(u8)]
+enum State {
+    /// No waiters are queued, and there is no pending notification.
+    /// Waiting while the queue is in this state will enqueue the waiter;
+    /// notifying while in this state will store a pending notification in the
+    /// queue, transitioning to [`State::Woken`].
+    Empty = 0b00,
+
+    /// There are one or more waiters in the queue. Waiting while
+    /// the queue is in this state will not transition the state. Waking while
+    /// in this state will wake the first waiter in the queue; if this empties
+    /// the queue, then the queue will transition to [`State::Empty`].
+    Waiting = 0b01,
+
+    /// The queue has a stored notification. Waiting while the queue
+    /// is in this state will consume the pending notification *without*
+    /// enqueueing the waiter and transition the queue to [`State::Empty`].
+    /// Waking while in this state will leave the queue in this state.
+    Woken = 0b10,
+
+    /// The queue is closed. Waiting while in this state will return
+    /// [`Closed`] without transitioning the queue's state.
+    ///
+    /// *Note*: This *must* correspond to all state bits being set, as it's set
+    /// via a [`fetch_or`].
+    ///
+    /// [`Closed`]: crate::wait::Closed
+    /// [`fetch_or`]: core::sync::atomic::AtomicUsize::fetch_or
+    Closed = 0b11,
+}
+
+#[derive(Clone, Debug)]
+enum Wakeup {
+    Empty,
+    Waiting(Waker),
+    One,
+    Closed,
+}
+
+// === impl DataWaitQueue ===
+
+impl<K: PartialEq, V> DataWaitQueue<K, V> {
+    /// Returns a new `DataWaitQueue`.
+    #[must_use]
+    #[cfg(not(loom))]
+    pub const fn new() -> Self {
+        Self {
+            state: CachePadded::new(AtomicUsize::new(State::Empty.into_usize())),
+            queue: Mutex::new(List::new()),
+            _key: PhantomData,
+            _val: PhantomData,
+        }
+    }
+
+    /// Returns a new `DataWaitQueue`.
+    #[must_use]
+    #[cfg(loom)]
+    pub fn new() -> Self {
+        Self {
+            state: CachePadded::new(AtomicUsize::new(State::Empty.into_usize())),
+            queue: Mutex::new(List::new()),
+            _key: PhantomData,
+            _val: PhantomData,
+        }
+    }
+
+    /// Wake a certain task in the queue.
+    ///
+    /// If the queue is empty, a wakeup is stored in the `DataWaitQueue`, and the
+    /// next call to [`wait`] will complete immediately.
+    ///
+    /// [`wait`]: DataWaitQueue::wait
+    #[inline]
+    pub fn wake(&self, key: &K, val: V) -> WakeOutcome<V> {
+        // snapshot the queue's current state.
+        let mut state = self.load();
+
+        // check if any tasks are currently waiting on this queue. if there are
+        // no waiting tasks, store the wakeup to be consumed by the next call to
+        // `wait`.
+        loop {
+            match state.get(QueueState::STATE) {
+                // if the queue is closed, bail.
+                State::Closed => return WakeOutcome::NoMatch(val),
```

Comment:
is this really `NoMatch`? perhaps there should be an additional `WakeOutcome` variant for "queue is already closed"?

Reply:
No objections! "NoMatch" really means "I didn't wake anyone, here's your data back", but might be worth distinguishing.